### PR TITLE
[sk] Added syntax error cleaning suggestion

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
Added the cleaning suggestion for fixing syntax errors. Some notable points
- Adjusted the syntax error check to exclude null value placeholders as well (i.e. the placeholder `'missing'` should not be counted as invalid)
- Refactored `build_action_variables` to be a shared function of both `impute_values` and `fix_syntax_error`, moved to `data_cleaner/cleaning_rules/shared.py`
- Added the cleaning suggestion to the base pipeline.

# Tests
Unit tests were written to confirm that the cleaning suggestion is properly applied and generates the correct payload. In addition, testing was performed on user datasets to confirm that syntax error problems are generated and fixed correctly.

cc:
@wangxiaoyou1993

### CODE REVIEW

File: mage_ai/data_cleaner/cleaning_rules/fix_syntax_errors.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        for column in self.df_columns:
+            if self.statistics[f'{column}/invalid_value_rate'] > 0:
+                columns_with_syntax_errors.append(column)
+
+        suggestions = []
+        if len(columns_with_syntax_errors):
+            suggestions.append(
+                self._build_transformer_action_suggestion(
+                    title='Fix Syntax Errors',
+                    message='Fix syntactical errors to reduce the amount of noise in the data.',
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/cleaning_rules/fix_syntax_errors.py:21]: nit: "Fix syntax errors"

File: mage_ai/data_cleaner/cleaning_rules/shared.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+def build_action_variables(columns, column_types):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/cleaning_rules/shared.py:1]: You can just put it in BaseRule if you only use it in cleaning rules.

File: mage_ai/data_cleaner/column_types/column_type_detector.py [hunk 3]
```diff
@@ -38,34 +42,46 @@ def find_syntax_errors(series, column_type):
...
         return pd.Series([])
     dtype = series.dtype
-    str_series = series
+    invalid_placeholder = INVALID_VALUE_PLACEHOLDERS[column_type]
+    null_placeholder = CONSTANT_IMPUTATION_DEFAULTS[column_type]
+    mask = pd.Series([False] * len(series))
+    mask.index = series.index
     if column_type == ColumnType.EMAIL:
-        pattern = REGEX_EMAIL
+        placeholder_matches = series.str.match(invalid_placeholder, na=True)
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_types/column_type_detector.py:50]: Could you reduce duplicate code in this method?

File: mage_ai/data_cleaner/cleaning_rules/fix_syntax_errors.py (hunk 1 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/cleaning_rules/fix_syntax_errors.py:20]: Fixed

File: mage_ai/data_cleaner/cleaning_rules/shared.py (hunk 2 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/cleaning_rules/shared.py]: Forgot about this, I moved this function into BaseRule

File: mage_ai/data_cleaner/column_types/column_type_detector.py (hunk 3 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_types/column_type_detector.py]: Refactored to reduce duplicated code + added a new test case to make sure that the added logic works correctly.

File: mage_ai/data_cleaner/cleaning_rules/base.py [hunk 4]
```diff
@@ -42,3 +43,15 @@ def _build_transformer_action_suggestion(
                 outputs=outputs,
             ),
         )
+
+    def build_action_variables(self, columns):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/cleaning_rules/base.py:47]: You can make it a protected method.

File: mage_ai/data_cleaner/column_types/column_type_detector.py [hunk 5]
```diff
@@ -60,12 +70,21 @@ def find_syntax_errors(series, column_type):
         and not dtype is pd.Timestamp
     ):
         str_series = str_series.astype(str)
+        filter_invalid = True
```
Comment (wangxiaoyou1993) [inline mage_ai/data_cleaner/column_types/column_type_detector.py:73]: actually, for invalid datetime, let's fill in with NaT

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/data_cleaner/cleaning_rules/base.py (hunk 4 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/cleaning_rules/base.py]: Made this method protected

File: mage_ai/data_cleaner/column_types/column_type_detector.py (hunk 5 above)
Comment (skunichetty) [reply mage_ai/data_cleaner/column_types/column_type_detector.py]: Invalid datetimes now resolve to NaT

**Pull request merged**